### DISCUSSION TITLE
Use without bootstrap-icons

### DISCUSSION THREAD

**Discussion by xvaara** [cdmoro/bootstrap-vue-3#179] [Ideas]
I'd be great to be able to use without bootstrap-icons, since unplugin-icons with vite (or other bundlers) is so good at only including used icons.

**Comment by jclijmans** [comment]
I'm not happy with the current implementation either. it loads all icons in whether they are used or not.
Further it only works internally with svg sprites.

But, they currently work and for now that's ok by me.

For the future : Can we use unplugin-icons underneath and still keep the current api interface like the BIcon component with icon property?

<b-icon icon="arrow-up"></b-icon>

I think we need to create a icon component for each known icon which then does the import of the unplugin-icons.
That's a lot of work and difficult to maintain. Each new known icon must be added.

Just my 2 cents

**Comment by cdmoro** [comment]
I agree we shouldn't import all icons, is pointless. I like the way that BootstrapVue team implemented it. They run a script that generates all the icons, I don't know the cost of replicating that behavior in our library.

**Comment by zoranpopovic** [comment]
unplugin-icons really shines when you use it with auto-import: https://github.com/antfu/unplugin-icons#auto-importing 
After using that setup, which works in vue2 and vue3, it's hard to think that BootstrapVue's or any other approach makes sense, especially considering that all collections  (not just BS) are just one kebab string away from being in your app.  For this reason, I imagine that bootstrap-icons are at the very bottom of the list of things people want out of this library.

**Comment by xvaara** [comment]
I think we should remove it from the default bundle, export it as it's own tree shakable component. 
Or just remove it since I think most people will move to automatic icons... unplugin-icons uses iconify project that even has quite awesome component for vue3: https://github.com/iconify/iconify/tree/master/packages/vue 

Can we offer anything close to that? Maybe just use that component as basis if one is needed?
we could use their json to generate the icons dynamically at build time: https://github.com/iconify/collections-json/blob/master/json/bi.json

After all these options, I'd say just drop it. Less is More. Document people to use unplugin or iconify component.

**Comment by jclijmans** [comment]
I don't think we can just remove the icons from the package.
Lots of existing applications rely on b-icon being there
Also some components like BAvatar use BIcon internally

**Comment by cdmoro** [comment]
I think we have two scenarios, replicate BootstrapVue behavior or drop Bootstrap's icons. I agree with @zoranpopovic that bootstrap-icons is the last thing that our users want about this library.

I would like that more people or at least the contributors vote to determinate the next step.

**Comment by jclijmans** [comment]
It should be put on the roadmap though  because it is a real issue that should be solved.

Ideally we should still have the b-icon component but it uses dynamically imported icons underneath.

**Comment by jclijmans** [comment]
I found a solution!!

Let me first say that I like to keep the BIcon component because it provides icon manipulation out-of-the-box like color variant, sizing, svg transforms (flip, shift, stacking)
I don't care about the specific icon-set.

So I started testing with the unplugin-icons system and came up with a solution to use it, so we don't need to include the complete BootstrapIcon set.

I tested it out with the bootstrap-icons set and with the material-design icon set. All examples from the docs work!!

- Install unplugin-icons : yarn add --dev unplugin-icons
- Update the vite.config.ts file to include the Icons option [see](https://github.com/antfu/unplugin-icons)
- Install the icon sets because I dont want the whole set
    - bootstrap-icons : yarn add --dev @iconify-json/bi
    - material design : yarn add --dev @iconify-json/mdi

I updated BIcon to handle the icon svg's

### Usage

- You must import the icons you want to use (maybe auto import will work eventually, but I'm not a big fan)
- You must the define a variable to hold the icon
Now you can use the icon in a BICon component using the :icon binding

### Example

```
<template>
  <b-card>
    <div>
      <p class="h1 mb-2">Icon <b-icon :icon="cartVariant"></b-icon2></p>
      <p class="h2 mb-2">Icon <b-icon :icon="camera"></b-icon2></p>
      <p class="h3 mb-2">Icon <b-icon :icon="cartVariant"></b-icon2></p>
      <p class="h4 mb-2">Icon <b-icon :icon="camera"></b-icon2></p>
      <p class="h5 mb-2">Icon <b-icon :icon="cartVariant"></b-icon2></p>
    </div>
  </b-card>
</template>

<script lang="ts" setup>
  import Camera from '~icons/bi/camera'
  import CartVariant from '~icons/mdi/cart-variant'

  const camera = Camera
  const cartVariant = CartVariant
</script>
```

### Result

![image](https://user-images.githubusercontent.com/2701545/138596667-202a30ec-0e35-4ebb-8e2f-ea442a844d72.png)

**Reply by cdmoro** [reply]
I like this solution!

**Comment by jclijmans** [comment]
Any conclusion whether we go forward with this solution?

**Comment by zoranpopovic** [comment]
it's great that it works.  the extra variable creation is a bit cumbersome.  it cannot be avoided?  maybe with import as? i always thought that import is similar to const var assignment.  does var need to be reactive?  Otherwise, the approach needs good documentation.

also, is iconify then an optional dependency to bv3?

**Reply by zoranpopovic** [reply]
also, looking at iconify's vue component has all the flips, rescaling. https://docs.iconify.design/icon-components/vue/
does b-icon have an even fancier set that is missing in iconify?

**Comment by jclijmans** [comment]
The approach here uses unplugin-icons. This uses iconify if I'm correct.
This means that bootstrap-vue-3 indeed has a dependency on unplug-icons 

The approach imports SVG's , so I think it could work with every icon SVG that can be imported.

The reason an SVG Icon has to be imported is that the BIcon uses the component :is="..." approach. As stated in many discussions over at the github repo for unplug-icons, the icons cannot be dynamically loaded, that would bypass the tree-shaking.

I like unplugin -icons because you can install only the icon set that you want. I don't want the whole 100+ icon sets in my node_modules folder.
Unplug-icons supports tree-shaking, so only used icons are present in the deployed app folder. Don't know if iconify supports tree-shaking.

Regarding the transformations : is it correct that iconify's vue component only supports flip and rotation? and rotation only by 90, 180, 270 degrees?
BIcon currently supports flip, rotation (any degree), shift and stacking

**Reply by jclijmans** [reply]
To further explain :
With this approach, BICon and Boostrap-vue-3 have no direct dependency on unplugin-icons.
Any icon set that supports importing individual icons as svg is supported.

It's up to the user to install unplugin-icons in their app if they want to use BIcon

**Comment by xvaara** [comment]
Added the ability to skip b-icon (and b-avatar since it imports b-icon) in #222 
Not sure if it treeshakes them out of the bundle.

**Reply by jclijmans** [reply]
I don't think tree shaking will work with your change in #222 because the remove is only applied at runtime.
The compiler/bundler doesn't know anything about it, so cannot tree-shake the removed components.

**Reply by cdmoro** [reply]
It doesn't affect the bundle size at all. Because when you use this library as a plugin you always import the whole js, It only avoids the component (b-avatar and b-icon in your example) being declared globally. The solution for that is not to use it as a plugin and import every component individually.

**Comment by mariusa** [comment]
In order to KISS, I'll remove bootstrap icons from this project. See https://github.com/cdmoro/bootstrap-vue-3/issues/255 for more info.
If one wants to *optionally* use bootstrap icons, one can do it similarly with any other icon font `<i class="..."></i>`. If some people prefer `b-icon`, then please make a new package, eg `bootstrap-icons-vue`, which interested people can install like any other library.
